### PR TITLE
(BKR-242) Docker swarm

### PR DESCRIPTION
**Pull request by electrical**
Several fixes to be able to use Swarm with beaker.

When we want the IP address of the container with running swarm we have to fetch the IP of the swarm slave it self.
This is required since DOCKER_HOST env var will point to the Swarm Master and not the resulting docker host where the image will be running.

By default the build command is ran on a random node, the Container create + run command could be executed on a completely different node causing the create to fail since that node does not have just built image.

To distribute the same built image between all swarm slaves we need to push the image to a private registry.
If the image does not exist yet ( first build ) we tag + push it, If it does exist we will re-use it.
If we built the image on Node A but then the run command is executed on
node B we will pull the image automatically and run it.

The only downside is that we will always build the image since we need to know the resulting image ID.
Hopefully this can be improved in the future.

### CODE REVIEW

File: lib/beaker/hypervisor/docker.rb [hunk 1]
```diff
@@ -21,6 +21,15 @@ def initialize(hosts, options)
 
       # Pass on all the logging from docker-api to the beaker logger instance
       ::Docker.logger = @logger
+
+      if ::Docker.version['Version'] =~ /swarm/
+        @docker_type = 'swarm'
+        raise "Using Swarm with beaker requires a private regitry. Please setup the private registry and set the 'DOCKER_REGISTRY' env var" unless ENV['DOCKER_REGISTRY']
```
Comment (richardc) [inline lib/beaker/hypervisor/docker.rb]: This really hides the adverbial unless.

Comment (electrical) [inline lib/beaker/hypervisor/docker.rb]: Sorry, how do you mean?

Comment (richardc) [inline lib/beaker/hypervisor/docker.rb]: You have to scroll a long way to the right to see that it's there.  It doesn't even show up in the diff view on github, which changes it from a long error message to a surprise condition.

Comment (richardc) [inline lib/beaker/hypervisor/docker.rb]: Structured like this;

```
unless ENV['DOCKER_REGISTRY']
  raise "Using Swarm with beaker requires a private regitry. Please setup the private registry and set the 'DOCKER_REGISTRY' env var"
end
```

You can see the condition, and it really is just a long string.

Comment (electrical) [inline lib/beaker/hypervisor/docker.rb]: Ahhh i see what you mean. will modify.

File: lib/beaker/hypervisor/docker.rb [hunk 2]
```diff
@@ -21,6 +21,18 @@ def initialize(hosts, options)
 
       # Pass on all the logging from docker-api to the beaker logger instance
       ::Docker.logger = @logger
+
+      if ::Docker.version['Version'] =~ /swarm/
+        @docker_type = 'swarm'
+        unless ENV['DOCKER_REGISTRY']
+          raise "Using Swarm with beaker requires a private regitry. Please setup the private registry and set the 'DOCKER_REGISTRY' env var"
```
Comment (fiddyspence) [inline lib/beaker/hypervisor/docker.rb]: 'regitry' is a typo

Comment (electrical) [inline lib/beaker/hypervisor/docker.rb]: thanks for catching that. fixing.

**Comment by electrical** [comment]
@kylog good points will process them. thanks!

**Comment by electrical** [comment]
@kylog i've implemented some changes based on your feedback.

File: lib/beaker/hypervisor/docker.rb [hunk 3]
```diff
@@ -21,6 +21,19 @@ def initialize(hosts, options)
 
       # Pass on all the logging from docker-api to the beaker logger instance
       ::Docker.logger = @logger
+
+      # Find out what kind of remote instance we are talking against
+      if ::Docker.version['Version'] =~ /swarm/
+        @docker_type = 'swarm'
+        unless ENV['DOCKER_REGISTRY']
+          raise "Using Swarm with beaker requires a private registry. Please setup the private registry and set the 'DOCKER_REGISTRY' env var"
```
Comment (anodelman) [inline lib/beaker/hypervisor/docker.rb]: Could you collect the list of supported docker env vars and put them into the initialize yard doc?

Comment (electrical) [inline lib/beaker/hypervisor/docker.rb]: will do

Comment (electrical) [inline lib/beaker/hypervisor/docker.rb]: Added some documentation, not sure if its in the form you want it. Couldn't find an other nice way to document ENV variables.

**Comment by anodelman** [comment]
retest this please

**Comment by electrical** [comment]
@anodelman Any idea when we can merge this? thanks!

**Comment by anodelman** [comment]
retest this please

**Comment by anodelman** [comment]
Errors during yard doc generation.

**Comment by electrical** [comment]
@anodelman fixed it. I really need to make a habit of running tests. sorry :-)

**Comment by anodelman** [comment]
retest this please

**Comment by electrical** [comment]
still failed :-(

**Comment by anodelman** [comment]
```
Failures:

  1) Beaker::Docker#initialize should require the docker gem
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:94:in `block (3 levels) in <module:Beaker>'

  2) Beaker::Docker#initialize should set Docker options
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:105:in `block (3 levels) in <module:Beaker>'

  3) Beaker::Docker#initialize should check the Docker gem can work with the api
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:121:in `block (3 levels) in <module:Beaker>'

  4) Beaker::Docker#initialize should hook the Beaker logger into the Docker one
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:127:in `block (3 levels) in <module:Beaker>'

  5) Beaker::Docker#initialize when Docker options are already set should not override Docker options
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:114:in `block (4 levels) in <module:Beaker>'

  6) Beaker::Docker#provision should call dockerfile_for with all the hosts
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:134:in `block (3 levels) in <module:Beaker>'

  7) Beaker::Docker#provision should pass the Dockerfile on to Docker::Image.create
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:134:in `block (3 levels) in <module:Beaker>'

  8) Beaker::Docker#provision should create a container based on the Image (identified by image.id)
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:134:in `block (3 levels) in <module:Beaker>'

  9) Beaker::Docker#provision should start the container
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:134:in `block (3 levels) in <module:Beaker>'

  10) Beaker::Docker#provision should generate a new /etc/hosts file referencing each host
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:134:in `block (3 levels) in <module:Beaker>'

  11) Beaker::Docker#provision should record the image and container for later
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:134:in `block (3 levels) in <module:Beaker>'

  12) Beaker::Docker#provision connecting to ssh should expose port 22 to beaker
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:134:in `block (3 levels) in <module:Beaker>'

  13) Beaker::Docker#provision connecting to ssh should expose port 22 to beaker when using DOCKER_HOST
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:134:in `block (3 levels) in <module:Beaker>'

  14) Beaker::Docker#provision connecting to ssh should have ssh agent forwarding enabled
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:134:in `block (3 levels) in <module:Beaker>'

  15) Beaker::Docker#cleanup should stop the containers
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:223:in `block (3 levels) in <module:Beaker>'

  16) Beaker::Docker#cleanup should delete the containers
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:223:in `block (3 levels) in <module:Beaker>'

  17) Beaker::Docker#cleanup should delete the images
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:223:in `block (3 levels) in <module:Beaker>'

  18) Beaker::Docker#cleanup should not delete the image if docker_preserve_image is set to true
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:223:in `block (3 levels) in <module:Beaker>'

  19) Beaker::Docker#cleanup should delete the image if docker_preserve_image is set to false
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:223:in `block (3 levels) in <module:Beaker>'

  20) Beaker::Docker#dockerfile_for should raise on an unsupported platform
     Failure/Error: expect { docker.send(:dockerfile_for, {'platform' => 'a_sidewalk' }) }.to raise_error(/platform a_sidewalk not yet supported on docker/)
       expected Exception with message matching /platform a_sidewalk not yet supported on docker/, got #<NoMethodError: undefined method `version' for Docker:Module> with backtrace:
         # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
         # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
         # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
         # ./spec/beaker/hypervisor/docker_spec.rb:270:in `block (4 levels) in <module:Beaker>'
         # ./spec/beaker/hypervisor/docker_spec.rb:270:in `block (3 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:270:in `block (3 levels) in <module:Beaker>'

  21) Beaker::Docker#dockerfile_for should add docker_image_commands as RUN statements
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:274:in `block (3 levels) in <module:Beaker>'

  22) Beaker::Docker#dockerfile_for should add docker_image_entrypoint
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
rake aborted!
Failed to 'bundle exec rspec' (exit status: pid 25049 exit 1)
/var/lib/jenkins/workspace/qe_beaker_unit-rb_rspec/ruby/1.9.3-p484/Rakefile:39:in `block (3 levels) in <top (required)>'
/var/lib/jenkins/workspace/qe_beaker_unit-rb_rspec/ruby/1.9.3-p484/Rakefile:33:in `block (2 levels) in <top (required)>'
/var/lib/jenkins/workspace/qe_beaker_unit-rb_rspec/ruby/1.9.3-p484/Rakefile:10:in `block in <top (required)>'
/var/lib/jenkins/workspace/qe_beaker_unit-rb_rspec/ruby/1.9.3-p484/Rakefile:23:in `block in <top (required)>'
/usr/local/rvm/gems/ruby-1.9.3-p484/bin/ruby_executable_hooks:15:in `eval'
/usr/local/rvm/gems/ruby-1.9.3-p484/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => test:spec
(See full trace by running task with --trace)
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:287:in `block (3 levels) in <module:Beaker>'

  23) Beaker::Docker#dockerfile_for should use zypper on sles
     Failure/Error: let (:docker) { ::Beaker::Docker.new( hosts, options ) }
     NoMethodError:
       undefined method `version' for Docker:Module
     # ./lib/beaker/hypervisor/docker.rb:33:in `initialize'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `new'
     # ./spec/beaker/hypervisor/docker_spec.rb:60:in `block (2 levels) in <module:Beaker>'
     # ./spec/beaker/hypervisor/docker_spec.rb:296:in `block (3 levels) in <module:Beaker>'

Finished in 14.71 seconds (files took 2.07 seconds to load)
898 examples, 23 failures
```

**Comment by anodelman** [comment]
retest this please

**Pull request merged**